### PR TITLE
Gemma3 vision models keras to hf checkpoint conversion

### PR DESCRIPTION
**Pull request by kharshith-k**
Here's the [Colab Gist](https://colab.research.google.com/gist/kharshith-k/9fa98607683be051cdcd4d79ec8fb4e4/gemma3_multimodal_h5_to_safetensors_conversion_keras_hub.ipynb) link for Gemma3-Vision models conversion from keras to safetensors format

image

image

image

### CODE REVIEW

File: keras_hub/src/utils/transformers/export/hf_exporter.py [hunk 1]
```diff
@@ -142,15 +148,13 @@ def export_tokenizer(tokenizer, path):
...
-
     # Save tokenizer assets
     tokenizer.save_assets(path)
-
     # Export tokenizer config
     tokenizer_type = tokenizer.__class__.__name__
     if tokenizer_type not in MODEL_TOKENIZER_CONFIGS:
         raise ValueError(
-            f"Export to Transformer format not implemented for {tokenizer_type}"
+            "Export to Transformers format not implemented for {tokenizer_type}"
```
Comment (sachinprasadhs) [inline keras_hub/src/utils/transformers/export/hf_exporter.py]: Missing f string here

File: keras_hub/src/utils/transformers/export/gemma3.py [hunk 2]
```diff
@@ -37,6 +38,38 @@ def get_gemma3_config(backbone):
...
+            "image_size": image_encoder.image_size,
+            "patch_size": image_encoder.patch_size,
+            "num_attention_heads": image_encoder.num_heads,
+            "hidden_size": image_encoder.hidden_dim,
+            "num_hidden_layers": image_encoder.num_layers,
+            "intermediate_size": image_encoder.intermediate_dim,
+            "layer_norm_eps": image_encoder.layer_norm_epsilon,
+            "model_type": "siglip_vision_model",
+            "vision_use_head": False,
+        }
```
Comment (sachinprasadhs) [inline keras_hub/src/utils/transformers/export/gemma3.py:113]: Doesn't it need     
```
"boi_token_index": 255999,          
    "eoi_token_index": 256000,         
    "image_token_index": 262144,   
    "mm_tokens_per_image": 256, 
```

**Review by sachinprasadhs — COMMENTED** [review]
Please find my initial review.

**Review by sachinprasadhs — COMMENTED** [review]
Get rope_theta from config.
Basically, for all the config, follow how the configs are pulled from here https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/utils/transformers/convert_gemma3.py and do the reverse engineering.
This should validate for Translate Gemma and Embedding Gemma.

Also in the test it's still showing `vision_encoder=None` with #TODO, if you are doing vision test in another test, make the chanegs to the commenst accordingly.

**Comment by kharshith-k** [comment]
> Get rope_theta from config. Basically, for all the config, follow how the configs are pulled from here https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/utils/transformers/convert_gemma3.py and do the reverse engineering. This should validate for Translate Gemma and Embedding Gemma.
> 
> Also in the test it's still showing `vision_encoder=None` with #TODO, if you are doing vision test in another test, make the chanegs to the commenst accordingly.

Hi @sachinprasadhs , I've addressed above mentioned comments. Please review the changes. Thanks!

File: keras_hub/src/utils/transformers/export/gemma3.py [hunk 3]
```diff
@@ -11,9 +11,23 @@ def get_gemma3_config(backbone):
...
-    hf_config = {
-        "architectures": ["Gemma3ForCausalLM"],
-        "model_type": "gemma3_text",
+    # Check if this is a vision model
+    has_vision = backbone.vision_encoder is not None
+
+    # Extract rope_theta values from layers
+    # Sliding/local attention layers typically use 10K, full/global use 1M
+    rope_theta_local = 10000.0  # Default for sliding attention
+    rope_theta_global = 1000000.0  # Default for full attention
```
Comment (laxmareddyp) [inline keras_hub/src/utils/transformers/export/gemma3.py]: How did you arrive at these number ? Please document it with source?

File: keras_hub/src/utils/transformers/export/gemma3.py [hunk 4]
```diff
@@ -11,9 +11,23 @@ def get_gemma3_config(backbone):
...
+
+    for i in range(backbone.num_layers):
+        layer = backbone.get_layer(f"decoder_block_{i}")
+        if layer.use_sliding_window_attention:
+            rope_theta_local = layer.rope_wavelength
+        else:
+            rope_theta_global = layer.rope_wavelength
+
+    # Base text config
+    text_config = {
```
Comment (laxmareddyp) [inline keras_hub/src/utils/transformers/export/gemma3.py]: Why  did we remove these from text config?

https://github.com/keras-team/keras-hub/blob/d512b282922e38fc29d87476255fba7ea0d3ee73/keras_hub/src/utils/transformers/export/gemma3.py#L18-L20

File: keras_hub/src/utils/transformers/export/gemma3.py [hunk 5]
```diff
@@ -169,21 +402,61 @@ def get_gemma3_tokenizer_config(tokenizer):
...
+            # Only add special tokens (those with < >)
+            is_special = token.startswith("<") and token.endswith(">")
+            if is_special or token.startswith("<unused"):
+                added_tokens_decoder[str(token_id)] = {
+                    "content": token,
+                    "special": True,
+                    "single_word": False,
+                    "lstrip": False,
+                    "rstrip": False,
+                    "normalized": False,
```
Comment (laxmareddyp) [inline keras_hub/src/utils/transformers/export/gemma3.py]: This scans all vocab_size tokens (256K+ for Gemma3) to find special tokens. The old code just checked a known list of ~7 tokens. this calls id_to_token() 256K+ times during every export — noticeably slow.

Can we keep a curated list like the original?

If you need all tokens (not just the known ones), worth a comment explaining why the full scan is necessary.

File: keras_hub/src/utils/transformers/export/gemma3.py [hunk 6]
```diff
@@ -38,7 +51,66 @@ def get_gemma3_config(backbone):
...
+            "num_attention_heads": image_encoder.num_heads,
+            "hidden_size": image_encoder.hidden_dim,
+            "num_hidden_layers": image_encoder.num_layers,
+            "intermediate_size": image_encoder.intermediate_dim,
+            "layer_norm_eps": image_encoder.layer_norm_epsilon,
+            "model_type": "siglip_vision_model",
+            "vision_use_head": False,
+            "boi_token_index": 255999,
+            "eoi_token_index": 256000,
+            "image_token_index": 262144,
```
Comment (laxmareddyp) [inline keras_hub/src/utils/transformers/export/gemma3.py]: Why hardcoded, can we not get it from config?

File: keras_hub/src/utils/transformers/export/gemma3.py [hunk 7]
```diff
@@ -195,5 +468,111 @@ def get_gemma3_tokenizer_config(tokenizer):
...
+        "eos_token_id": 1,
+        "pad_token_id": 0,
+        "transformers_version": "4.47.1",
+    }
+
+    # Add soft capping if configured
+    if (
+        hasattr(backbone, "final_logit_soft_cap")
+        and backbone.final_logit_soft_cap
+    ):
```
Comment (laxmareddyp) [inline keras_hub/src/utils/transformers/export/gemma3.py]: nit: hasattr is unnecessary here — Gemma3Backbone.__init__ always sets self.final_logit_soft_cap. Can simplify to just if backbone.final_logit_soft_cap:. The hasattr guard masks potential bugs if the attribute is ever renamed.

File: keras_hub/src/utils/transformers/export/gemma3_test.py [hunk 8]
```diff
@@ -165,3 +171,309 @@ def test_export_to_hf(self):
...
+        keras_model.export_to_transformers(export_path)
+
+        # Verify processor config exists for vision models
+        processor_config_path = os.path.join(
+            export_path, "processor_config.json"
+        )
+        self.assertTrue(
+            os.path.exists(processor_config_path),
+            "processor_config.json should exist for vision models",
+        )
```
Comment (laxmareddyp) [inline keras_hub/src/utils/transformers/export/gemma3_test.py:251]: The vision export is the main feature of this PR but the test never verifies config.json. Can we add a few asserts after L230? like below : 

with open(os.path.join(export_path, "config.json"), "r") as f:
    config = json.load(f)
self.assertEqual(config["model_type"], "gemma3")
self.assertIn("text_config", config)
self.assertIn("vision_config", config)
self.assertEqual(config["vision_config"]["model_type"], "siglip_vision_model")

File: tools/checkpoint_export/verify_gemma3_export.py [hunk 9]
```diff
@@ -0,0 +1,727 @@
...
+            exp_model = AutoModelForImageTextToText.from_pretrained(
+                export_path, dtype=torch.float32
+            )
+            exp_processor = AutoProcessor.from_pretrained(hf_model_id)
+            exp_tokenizer = exp_processor.tokenizer
+        else:
+            exp_model = AutoModelForCausalLM.from_pretrained(
+                export_path, dtype=torch.float32
+            )
+            exp_tokenizer = AutoTokenizer.from_pretrained(hf_model_id)
```
Comment (laxmareddyp) [inline tools/checkpoint_export/verify_gemma3_export.py]: validate_exported_model loads tokenizer from hf_model_id (original HF repo), NOT from export_path. The exported tokenizer is never validated?

File: tools/checkpoint_export/verify_gemma3_export.py [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+            gen_out = hf_model.generate(
+                **hf_inputs, max_new_tokens=30, do_sample=False
+            )
+        results["text_generated"] = hf_tokenizer.decode(
+            gen_out[0][hf_inputs["input_ids"].shape[1] :],
+            skip_special_tokens=True,
+        )
+        print(f'    Text generation: "{results["text_generated"][:80]}"')
+
+    # TODO: Add vision validation when image processing is supported
```
Comment (laxmareddyp) [inline tools/checkpoint_export/verify_gemma3_export.py]: Vision logit validation TODO'd out?

**Review by laxmareddyp — CHANGES_REQUESTED** [review]
Thanks for the PR..!  Left few comments, please address it.

**Review by laxmareddyp — APPROVED** [review]
Thanks for the PR.!
